src/Microsoft.ML.Core/Data/IValueMapper.cs

```diff
@@ -9,13 +9,13 @@ namespace Microsoft.ML.Runtime.Data
     /// <summary>
     /// Delegate type to map/convert a value.
     /// </summary>
-    public delegate void ValueMapper<TSrc, TDst>(ref TSrc src, ref TDst dst);
+    public delegate void ValueMapper<TSrc, TDst>(in TSrc src, ref TDst dst);
```

Comment:
Just to confirm, is `TSrc` expected to always be a `readonly struct`? `in` causes hidden copies when calling an instance member for non-readonly structs and can cause unexpected perf-regressions in some cases.

Reply:
As a practical matter in *most* applications they will be some sort of `VBuffer`, with some isolated instances here and there as you've observed. Yet these other types from what I see tend to also be readonly structs. (E.g., you observed elsewhere [`TimeSpan`](https://github.com/dotnet/corefx/blob/48eaa759411a0d54f63c61d4aa51fa71482d53cc/src/Common/src/CoreLib/System/TimeSpan.cs#L31), and [here is `Int32`](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Int32.cs#L15)) I just ran a pretty simple benchmark like so (using BenchmarkDotNet, my new favorite project)...

```csharp
private static int Foo(in byte a, in byte b) => a + b;
private static int Bar(byte a, byte b) => a + b;

[Benchmark]
public void Foo() { for (int i = 0; i < 1000; ++i) Foo(2, 5); }
[Benchmark]
public void Bar() { for (int i = 0; i < 1000; ++i) Bar(2, 5); }
```

At least on my computer this gives:

| Method |     Mean |     Error |    StdDev |
|------- |---------:|----------:|----------:|
|    Foo | 239.6 ns | 2.8371 ns | 2.6538 ns |
|    Bar | 238.2 ns | 0.8780 ns | 0.7332 ns |

So it seems like it is at least not harmful for things like `int`, while still giving some benefit for things like `VBuffer`.

Reply:
In my limited knowledge of the area, I would guess that `>95%` of the time ValueMapper will be used for primitives, reference types, and `readonly struct`.

The most common case is to use these types:

```
    using BL = Boolean;
    using DT = DateTime;
    using DZ = DateTimeOffset;
    using R4 = Single;
    using R8 = Double;
    using I1 = SByte;
    using I2 = Int16;
    using I4 = Int32;
    using I8 = Int64;
    using SB = StringBuilder;
    using TX = ReadOnlyMemory<char>;
    using TS = TimeSpan;
    using U1 = Byte;
    using U2 = UInt16;
    using U4 = UInt32;
    using U8 = UInt64;
    using UG = UInt128;
```

Along with `VBuffer`, which just got made `readonly` in #1454.

Skimming this list, I found that `UInt128` was not marked `readonly`. So I'm opening #1496 to make that (and other obvious readonly structs) marked as `readonly`.

Reply:
Is there any concern with the primitive types not being `readonly structs` for full framework and/or netstandard right now?